### PR TITLE
[WIP] New plugin: UnderglowControl

### PR DESCRIPTION
**Pull request by algernon**
The bulk of this is the `UnderglowControl` plugin, modeled after `LEDControl`. The `KBD4x` keyboard is the first one to support this, and I ported a few of the existing LED effects over.

Documentation is missing, but creating a PR makes it easier to review naming, the API, etc. I'll write the docs once we're happy with how things look.

### CODE REVIEW

**Comment by mattvenn** [comment]
looks like a great start, thanks for your work on this.

File: src/kaleidoscope/hardware/kbdfans/KBD4x.h [hunk 1]
```diff
@@ -24,9 +24,19 @@
...
 #include "kaleidoscope/macro_helpers.h"
-#include "kaleidoscope/hardware/avr/pins_and_ports.h"
+
+struct cRGB {
+  uint8_t g;
+  uint8_t r;
+  uint8_t b;
+};
+
+#define CRGB(r, g, b) (cRGB){g, r, b}
```
Comment (obra) [inline src/kaleidoscope/hardware/kbdfans/KBD4x.h]: I feel like this should be a separate PR -- it's adding color support to the 4x, not really adding underglow to kaleidoscope

File: src/kaleidoscope/hardware/kbdfans/KBD4x.h [hunk 2]
```diff
@@ -24,9 +24,19 @@
...
 
 #include "kaleidoscope/hardware/ATMegaKeyboard.h"
+#include "kaleidoscope/driver/led/WS2812.h"
+
+using Color = kaleidoscope::driver::led::color::GRB;
```
Comment (obra) [inline src/kaleidoscope/hardware/kbdfans/KBD4x.h]: ...same as above. But also, now I wonder if we could add helpers for the standard color orders, to reduce the chance of implementation code drift. So that you could just declare the color order and get your CRGB, cRGB, etc. It might also make it easier when we have to do cRGBW...

File: src/kaleidoscope/plugin/IdleLEDs.cpp [hunk 3]
```diff
@@ -25,23 +26,33 @@ uint16_t IdleLEDs::idle_time_limit = 600; // 10 minutes
 uint32_t IdleLEDs::end_time_;
 
 EventHandlerResult IdleLEDs::beforeEachCycle() {
-  if (!::LEDControl.paused &&
-      (Kaleidoscope.millisAtCycleStart() >= end_time_)) {
+  if (Kaleidoscope.millisAtCycleStart() < end_time_)
+    return EventHandlerResult::OK;
+
+  if (::Kaleidoscope.has_leds && !::LEDControl.paused) {
```
Comment (obra) [inline src/kaleidoscope/plugin/IdleLEDs.cpp]: I think this is a small change that can land without the underglow PR

File: src/kaleidoscope/plugin/IdleLEDs.cpp [hunk 4]
```diff
@@ -25,23 +26,33 @@ uint16_t IdleLEDs::idle_time_limit = 600; // 10 minutes
...
     ::LEDControl.set_all_leds_to(CRGB(0, 0, 0));
     ::LEDControl.syncLeds();
 
     ::LEDControl.paused = true;
   }
+  if (::UnderglowControl.has_leds() && !::UnderglowControl.isPaused()) {
+    ::UnderglowControl.setColor(0, 0, 0);
```
Comment (obra) [inline src/kaleidoscope/plugin/IdleLEDs.cpp]: I feel weird that this is "setColor(0,0,0)" and the LEDControl version is set_all_leds_to(CRGB(0,0,0)).

(LEDControl should at least grow setAllLedsTo() and this API and that one should use the same name and calling convention)

File: src/kaleidoscope/plugin/IdleLEDs.cpp [hunk 5]
```diff
@@ -25,23 +26,33 @@ uint16_t IdleLEDs::idle_time_limit = 600; // 10 minutes
...
+    ::UnderglowControl.sync();
```
Comment (obra) [inline src/kaleidoscope/plugin/IdleLEDs.cpp]: If this is called `sync`, then the version in LEDControl should also be called `sync`

File: src/kaleidoscope/plugin/IdleLEDs.cpp [hunk 6]
```diff
@@ -25,23 +26,33 @@ uint16_t IdleLEDs::idle_time_limit = 600; // 10 minutes
...
+    ::UnderglowControl.pause();
```
Comment (obra) [inline src/kaleidoscope/plugin/IdleLEDs.cpp]: LEDControl and this should have parity. If this is going to be a method here (which I like), it should get added to LEDControl at the same time.

File: src/kaleidoscope/plugin/IdleLEDs.cpp [hunk 7]
```diff
@@ -25,23 +26,33 @@ uint16_t IdleLEDs::idle_time_limit = 600; // 10 minutes
...
 }
 
 EventHandlerResult IdleLEDs::onKeyswitchEvent(Key &mapped_key, byte row, byte col, uint8_t key_state) {
 
-  if (::LEDControl.paused) {
+  if (Kaleidoscope.has_leds && ::LEDControl.paused) {
     ::LEDControl.paused = false;
     ::LEDControl.refreshAll();
   }
+  if (::UnderglowControl.has_leds() && ::UnderglowControl.isPaused()) {
```
Comment (obra) [inline src/kaleidoscope/plugin/IdleLEDs.cpp]: has_leds() and isPaused() look different. Feels weird.

File: src/kaleidoscope/plugin/IdleLEDs.cpp [hunk 8]
```diff
@@ -25,23 +26,33 @@ uint16_t IdleLEDs::idle_time_limit = 600; // 10 minutes
...
+    ::UnderglowControl.resume();
```
Comment (obra) [inline src/kaleidoscope/plugin/IdleLEDs.cpp]: LEDControl should grow resume()

**Comment by obra** [comment]
I've left a bunch of small notes inline, but as I've read through the code, I've started to wonder if it would make more sense to have a universal concept of a 'bank' of LEDs, with LEDControl and UnderglowControl both inheriting from, implementing or using it. 

A bank of LEDs may or may not have its own driver. (In the underglow keyboard I was designing, underglow and key LEDS were driven by the same IC)

Effects would 'run on' a given bank of LEDs

It'd be kind of crazy to have duplicate effects and implementations for, say, under-key, underglow, and side-glow.

Happy to talk more about this

File: src/kaleidoscope/hardware/kbdfans/KBD4x.h (hunk 1 above)
Comment (algernon) [reply src/kaleidoscope/hardware/kbdfans/KBD4x.h]: The reason I opted to add color (well, underglow) support to kbd4x in the same PR is because UnderglowControl relies on the hardware parts to work, and it felt weird having a dependency without anything to satisfy it.

I can lift the KBD4x part out with little effort though, so I'll do that.

File: src/kaleidoscope/hardware/kbdfans/KBD4x.h (hunk 2 above)
Comment (algernon) [reply src/kaleidoscope/hardware/kbdfans/KBD4x.h]: Yup! I have a half-baked branch that does just that. I'd keep that separate, because that's a fairly invasive change.

File: src/kaleidoscope/plugin/IdleLEDs.cpp (hunk 4 above)
Comment (algernon) [reply src/kaleidoscope/plugin/IdleLEDs.cpp]: Agreed. My thinking here was to play with names, to come up with something better than what we have in `LEDControl` (and then transition `LEDControl` over to the new name too). I'm not really fond of `setCrgbAt` for example, because it has the typename in the name too.

Btw, `setColor(CRGB(0, 0, 0))` also works, and so does `set_all_leds_to(r, g, b)`.

File: src/kaleidoscope/plugin/IdleLEDs.cpp (hunk 6 above)
Comment (algernon) [reply src/kaleidoscope/plugin/IdleLEDs.cpp]: Ok. I didn't want to touch `LEDControl` yet, before we agree on naming & API, happy to add the new methods there!

File: src/kaleidoscope/plugin/IdleLEDs.cpp (hunk 7 above)
Comment (algernon) [reply src/kaleidoscope/plugin/IdleLEDs.cpp]: Good point. Should be `hasLeds()` I think, right?

**Comment by algernon** [comment]
My original thinking (#376, previously keyboardio/Kaleidoscope-LEDControl#15) was similar, but the other way around: a common API for both, instead of a common ancestor. A common ancestor makes more sense. I'll give that idea a try next time I'm working on Underglow.

**Comment by obra** [comment]
what would you think of instead of a common ancestor both just being instantiations of the same type of object with different data?

**Comment by algernon** [comment]
That's pretty much what I ended up with before I drifted off to sleep last night. Something along these lines:

```c++
class LEDBank : public kaleidoscope::Plugin {
 public:
  LEDBank();

  const int8_t ledCount();
  const boolt isPerKey();
  void sync();

  void setColorAt(uint8_t index, cRGB color);
  void setColorAt(uint8_t index, uint8_t r, uint8_t g, uint8_t b);
  void setColorAt(byte row, byte col, cRGB color);
  void setColorAt(byte row, byte col, uint8_t r, uint8_t g, uint8_t b);
  void setColor(cRGB color);
  void setColor(uint8_t r, uint8_t g, uint8_t b);
  cRGB getColorAt(uint8_t index);
  cRGB getColorAt(byte row, byte col);

  EventHandlerResult beforeReportingState(); 
};
```

Methods that take `row,col` would be no-ops for banks that aren't per-key (could use a better name for `.isPerKey()` though).

On top of this, we'd have something like:

```c++
class LEDControl : public kaleidoscope::Plugin {
 public:
  LEDControl();

  const int8_t bankCount();
  LEDBank &getBank(int8_t bank);
  void useBanks(LEDBank banks[]);
  void selectBank(int8_t bank);

  // setColorAt, getColorAt, setColor, isPerKey(), ledCount() and sync(), as above
};
```

The `setColorAt()`, etc methods would use the bank selected with `selectBank()` (defaulting to bank 0). Effects would grow a `.selectBank(int8_t index)` method, which would store the index within the effect, and when it needs to do something with the leds, it'd do `::LEDControl.getBank(bank_).setColorAt(...)`.

To make things easier, we could also have an enum of banks, defined by the hardware libs:

```c++
enum class LEDBankName {
  PERKEY,
  UNDERGLOW,
  SIDEGLOW,
  NOT_SUPPORTED
};
```

If a board doesn't support a particular group, it can define the group to have the same value as `NOT_SUPPORTED`. This means that we could do something like `::LEDControl.getBank(UNDERGLOW).setColorAt(...)` and be it almost a no-op if underglow is not supported.

File: src/kaleidoscope/plugin/IdleLEDs.cpp (hunk 3 above)
Comment (algernon) [reply src/kaleidoscope/plugin/IdleLEDs.cpp]: I started to extract this into a separate PR, but I decided not to for the time being. Reading back the discussion here, I have the feeling it will not be needed. The change is here to support `UnderglowControl`, but if we bake underglow control into `LEDControl`, then this is a needless exercise, at least in its current form.

**Comment by obra** [comment]
@algernon -  I believe this is obsolete. If not, please feel free to reopen.

**Comment by algernon** [comment]
It is obsolete. We'll need to figure out a decent way to do this, but that's best done in an issue first, PR only after we have an idea what we're doing.
